checkpoint/side_handler.go

```diff
@@ -32,6 +34,45 @@ func NewSideTxHandler(k Keeper, contractCaller helper.IContractCaller) hmTypes.S
 	}
 }
 
+// SideHandleMsgCheckpointAdjust handles MsgCheckpointAdjust message for external call
+func SideHandleMsgCheckpointAdjust(ctx sdk.Context, k Keeper, msg types.MsgCheckpointAdjust, contractCaller helper.IContractCaller) (result abci.ResponseDeliverSideTx) {
+	logger := k.Logger(ctx)
+	chainParams := k.ck.GetParams(ctx).ChainParams
+	params := k.GetParams(ctx)
+
+	checkpointBuffer, err := k.GetCheckpointFromBuffer(ctx)
+	if checkpointBuffer != nil {
+		logger.Error("checkpoint buffer", "error", err)
+		return common.ErrorSideTx(k.Codespace(), common.CodeCheckpointBuffer)
+	}
+
+	checkpointObj, err := k.GetCheckpointByNumber(ctx, msg.HeaderIndex)
+	if err != nil {
+		logger.Error("Unable to get checkpoint from db", "error", err)
+		return common.ErrorSideTx(k.Codespace(), common.CodeOldCheckpoint)
+	}
+
+	rootChainInstance, err := contractCaller.GetRootChainInstance(chainParams.RootChainAddress.EthAddress())
+	if err != nil {
+		logger.Error("Unable to fetch rootchain contract instance", "error", err)
+		return common.ErrorSideTx(k.Codespace(), common.CodeOldCheckpoint)
+	}
+
+	root, start, end, _, _, err := contractCaller.GetHeaderInfo(msg.HeaderIndex, rootChainInstance, params.ChildBlockInterval)
+	if err != nil {
+		logger.Error("Unable to fetch checkpoint from rootchain", "error", err, "checkpointNumber", msg.HeaderIndex)
+		return common.ErrorSideTx(k.Codespace(), common.CodeOldCheckpoint)
+	}
+
+	if checkpointObj.EndBlock == end && checkpointObj.StartBlock == start && bytes.Equal(checkpointObj.RootHash.Bytes(), root.Bytes()) {
+		logger.Error("Same Checkpoint in DB")
+		return common.ErrorSideTx(k.Codespace(), common.CodeOldCheckpoint)
+	}
```

Comment:
We have to check if `msg` has the same data as `root, start, end, _, _, err` because we will be saving `msg` in the post handler `PostHandleMsgCheckpointAdjust`